### PR TITLE
A4A: Implement Shopping cart state based on Session storage.

### PR DESCRIPTION
**Pull request by jkguidaven**
This pull request adds a shopping cart state based on session storage. Storing the cart state on session storage allows the UI to remember previously selected items. It is common practice to use session storage as a caching mechanism for shopping carts on the frontend because it only keeps the items within the browser's tab context.

https://github.com/Automattic/wp-calypso/assets/56598660/c3c4dea5-99cf-402e-894a-56c06e0c2935

Closes https://github.com/Automattic/jetpack-genesis/issues/287

## Proposed Changes
* The PR renamed the Issue License context to the Shopping Cart Context to make the code more readable. The Issue License context no longer matches the Marketplace, which uses the Shopping cart model. Note that the IssueLicense component will be renamed in a separate PR.
* The PR added a new `useShoppingCart` hook, which will contain Shopping cart logic. The hook uses Session storage to cache the state. With session storage, we only cache the state until the page is closed (Browser tab). This is a common practice when implementing a frontend-based Shopping cart.
 

## Testing Instructions
* Switch branch: `git checkout add/a4a/persisting-shopping-cart-state`.
* Start the server by running `yarn start-a8c-for-agencies`.
* Go to http://agencies.localhost:3000/marketplace
* Select some items and confirm it is added to the shopping cart.
* Refresh the page to confirm that the shopping cart remembers the selected items.
* Go to the Hosting page and confirm that the shopping cart shows the exact items selected on the Products page.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/a8c-for-agencies/sections/marketplace/hooks/use-shopping-cart.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import { ShoppingCartItem } from '../types';
+
+const SELECTED_ITEMS_SESSION_STORAGE_KEY = 'shopping-card-selected-items';
+
+export default function useShoppingCart() {
+	const [ selectedItems, setSelectedItems ] = useState< ShoppingCartItem[] >( [] );
+
+	const { data } = useProductsQuery( true );
+
+	useEffect( () => {
```
Comment (jkguidaven) [inline client/a8c-for-agencies/sections/marketplace/hooks/use-shopping-cart.ts:13]: This is the main logic that loads selected items from cache.

File: client/a8c-for-agencies/sections/marketplace/context.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { createContext } from 'react';
+import type { ShoppingCartContext as ShoppingCartContextInterface } from './types';
+
+export const ShoppingCartContext = createContext< ShoppingCartContextInterface >( {
+	selectedItems: [],
```
Comment (yashwin) [inline client/a8c-for-agencies/sections/marketplace/context.ts]: `Items` is too generic. Can we name it as `CartItems`?

File: client/a8c-for-agencies/sections/marketplace/hosting-overview/index.tsx [hunk 3]
```diff
@@ -26,8 +30,17 @@ export default function Hosting() {
...
 					<Actions>
-						{ /* FIXME: We will need to have shared state for shopping cart items */ }
-						<ShoppingCart items={ [] } onRemoveItem={ () => {} } onCheckout={ () => {} } />
+						<ShoppingCart
+							items={ selectedItems }
+							onRemoveItem={ ( item: ShoppingCartItem ) => {
+								setSelectedItems(
+									selectedItems.filter( ( selectedItem ) => selectedItem !== item )
+								);
+							} }
```
Comment (yashwin) [inline client/a8c-for-agencies/sections/marketplace/hosting-overview/index.tsx]: I think we can move the remove item logic to the useShoppingCart hook

File: client/a8c-for-agencies/sections/marketplace/issue-license/licenses-form/index.tsx [hunk 4]
```diff
@@ -14,11 +14,11 @@ import {
 import useProductAndPlans from 'calypso/jetpack-cloud/sections/partner-portal/primary/issue-license/licenses-form/hooks/use-product-and-plans';
 import { useDispatch } from 'calypso/state';
 import { recordTracksEvent } from 'calypso/state/analytics/actions';
-import IssueLicenseContext from '../context';
+import { ShoppingCartContext } from '../../context';
+import { ShoppingCartItem } from '../../types';
```
Comment (yashwin) [inline client/a8c-for-agencies/sections/marketplace/issue-license/licenses-form/index.tsx]:
```suggestion
import type { ShoppingCartItem } from '../../types';
```

I think we can do this in all the files

**Review by yashwin — COMMENTED** [review]
It works as expected. I have left a few comments. Let me know your thoughts. Good job on this, @jkguidaven!

File: client/a8c-for-agencies/sections/marketplace/context.ts (hunk 2 above)
Comment (jkguidaven) [reply client/a8c-for-agencies/sections/marketplace/context.ts]: Good point 👍

File: client/a8c-for-agencies/sections/marketplace/hosting-overview/index.tsx (hunk 3 above)
Comment (jkguidaven) [reply client/a8c-for-agencies/sections/marketplace/hosting-overview/index.tsx]: Nice suggestion. This has been applied 👍

**Comment by jkguidaven** [comment]
@yashwin, thanks for the review. I made some adjustments to address your comments. Can you have another look? 🙏

**Review by yashwin — APPROVED** [review]
Thanks for addressing my comments.

LGTM now 👍

**Pull request merged**